category_encoders/woe.py

```diff
@@ -0,0 +1,290 @@
+    sigma: float
+        standard deviation (spread or "width") of the normal distribution.
+    regularization: float
+        Laplace correction. Must be strictly bigger than 1.
+
+    Example
+    -------
+    >>> from category_encoders import *
+    >>> import pandas as pd
+    >>> from sklearn.datasets import load_boston
+    >>> bunch = load_boston()
+    >>> y = bunch.target > 22.5
+    >>> X = pd.DataFrame(bunch.data, columns=bunch.feature_names)
+    >>> enc = WoeEncoder(cols=['CHAS', 'RAD']).fit(X, y)
+    >>> numeric_dataset = enc.transform(X)
+    >>> print(numeric_dataset.info())
+    <class 'pandas.core.frame.DataFrame'>
+    RangeIndex: 506 entries, 0 to 505
+    Data columns (total 13 columns):
+    CRIM       506 non-null float64
+    ZN         506 non-null float64
+    INDUS      506 non-null float64
+    NOX        506 non-null float64
+    RM         506 non-null float64
+    AGE        506 non-null float64
+    DIS        506 non-null float64
+    TAX        506 non-null float64
+    PTRATIO    506 non-null float64
+    B          506 non-null float64
+    LSTAT      506 non-null float64
+    CHAS       506 non-null float64
+    RAD        506 non-null float64
+    dtypes: float64(13)
+    memory usage: 51.5 KB
+    None
+
+    References
+    ----------
+
+    .. [1] Weight of Evidence (WOE) and Information Value Explained. from
+    https://www.listendata.com/2015/03/weight-of-evidence-woe-and-information.html.
+
+
+    """
+
+    def __init__(self, verbose=0, cols=None, drop_invariant=False, return_df=True, impute_missing=True,
+                 handle_unknown='impute', random_state=None, randomized=False, sigma=0.05, regularization=1.1):
+        self.verbose = verbose
+        self.return_df = return_df
+        self.drop_invariant = drop_invariant
+        self.drop_cols = []
+        self.cols = cols
+        self._dim = None
+        self.mapping = None
+        self.impute_missing = impute_missing
+        self.handle_unknown = handle_unknown
+        self._sum = None
+        self._count = None
+        self.random_state = random_state
+        self.randomized = randomized
+        self.sigma = sigma
+        self.regularization = regularization
+
+    # noinspection PyUnusedLocal
+    def fit(self, X, y, **kwargs):
+        """Fit encoder according to X and binary y.
+
+        Parameters
+        ----------
+
+        X : array-like, shape = [n_samples, n_features]
+            Training vectors, where n_samples is the number of samples
+            and n_features is the number of features.
+        y : array-like, shape = [n_samples]
+            Binary target values.
+
+        Returns
+        -------
+
+        self : encoder
+            Returns self.
+
+        """
+
+        # Unite parameters into pandas types
+        X = convert_input(X)
+        y = pd.Series(y, name='target')
+
+        # The lengths must be equal
+        assert X.shape[0] == y.shape[0]
```

Comment:
So this is another case of pretty subjective note: since asserts can be turned off globally by users, I tend to assume that they will be. So if it's important to check something, and its not internally invariant (these are checking user input), we should check them without an assert, and raise a value error. Otherwise we can't assume that these checks ever run.

Reply:
Agreed. Changed all asserts into exceptions.